### PR TITLE
Initial Setup

### PR DESCRIPTION
**Pull request by codenirvana**

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+    /**
+     * A function to flatten all items in the members array
+     *
+     * @param {Array} members - All members of an Item or ItemGroup
+     * @param {String} [path] - Path leading to members
+     */
+    flattenItems = function (members, path = []) {
+        var is = {};
+
+        members.forEach((element) => {
```
Comment (saswatds) [inline lib/index.js]: We should check that `members` is defined and an array

File: lib/index.js [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+                id: data.response.id,
+                status: data.response.status,
+                code: data.response.code,
+                responseTime: data.response.responseTime,
+                responseSize: data.response.responseSize
+            };
+        });
+
+        newman.on('assertion', function (err, data) {
+            var executionId = uuid(),
```
Comment (saswatds) [inline lib/index.js]: Need a comment explaining why errors are not being handled here

File: lib/index.js [hunk 3]
```diff
@@ -34,22 +34,20 @@ const
      * @param {Array} element - All members of an Item or ItemGroup
      * @param {String} [path] - Path leading to members
      */
-    flattenEvents = function (element, path = []) {
+    flattenEvents = function (element = [], path = []) {
+        if (!(element.events && element.items)) { return {}; }
```
Comment (saswatds) [inline lib/index.js]: This logic seems odd. What if event is present but item is missing, then we would want to add the events and skip the items

File: lib/index.js [hunk 4]
```diff
@@ -34,22 +34,20 @@ const
...
+
         var is = {};
 
         // Get the event and add it
-        if (element.events) {
-            element.events.members.forEach((event) => {
-                is[event.script.id] = { path };
-            });
-        }
+        element.events.members.forEach((event) => {
```
Comment (saswatds) [inline lib/index.js]: Type safety of members property is also a concern

File: lib/index.js [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+        return is;
+    },
+
+    /**
+     * A function to flatten all events in collection
+     *
+     * @param {Array} element - All members of an Item or ItemGroup
+     * @param {String} [path] - Path leading to members
+     */
+    flattenEvents = function (element = [], path = []) {
```
Comment (saswatds) [inline lib/index.js]: Element here is an object by default

File: lib/index.js [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+        if (element.events && element.events.members) {
+            element.events.members.forEach((event) => {
+                is[event.script.id] = { path };
+            });
+        }
+
+        // Traverse across all items and move through them
+        if (element.items && element.items.members) {
+            element.items.members.forEach((member) => {
+                Object.assign(is, flattenEvents(member, [...path, member.id]));
```
Comment (kunagpal) [inline lib/index.js]: Confirm that this works on Node v6.

File: lib/index.js [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+            typeof options.api === 'string' && (options.api = JSON.parse(options.api));
+        }
+        catch (e) {
+            return console.error('newman-reporter-remote:\n  error: ', e);
+        }
+
+        const runId = uuid(),
+            collectionInfo = cr.collection.toJSON().info;
+
+        var collection = {
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+     * @param {String} options.agentVersion - Remote agent version
+     * @param {Object} cr - The set of all the collection run options.
+     * @returns {*}
+     */
+    remoteReporter = function (newman, options, cr) {
+        if (!options.api) {
+            return console.error('newman-reporter-remote:\n  error: missing required option `api`');
+        }
+
+        try {
```
Comment (kunagpal) [inline lib/index.js]:
```suggestion
    if (typeof options.api === STRING) {
        try {
            options.api = JSON.parse(options.api);
        }
        catch (e) {
            return console.error('newman-reporter-remote:\n  error: ', e);
        }
    }
```

File: lib/index.js [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+            iterationStartTime,
+            itemStartTime,
+            requestStartTime,
+            consoles = {},
+
+            _events = flattenEvents(cr.collection),
+            _items = flattenItems(cr.collection.items.members);
+
+        newman.on('beforeIteration', function () {
+            iterationStartTime = Date.now();
```
Comment (kunagpal) [inline lib/index.js]: Add a comment explaining that this is intentional.

File: lib/index.js [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+        });
+
+        newman.on('iteration', function (err, data) {
+            if (err) {
+                !(err instanceof SerialisedError) && (err = new SerialisedError(err, true));
```
Comment (kunagpal) [inline lib/index.js]: Don't change the type of an entity on the fly. This leads to an optimization bailout (Link TBA)


```suggestion
            errors[err.id] = err instanceof SerialisedError ? err : new SerialisedError(err, true);
```

File: lib/index.js [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+                errors[err.id] = err;
+            }
+
+            const executionId = uuid();
+
+            executions.push({
+                id: executionId,
+                index: executions.length,
+                branch: 0,
+                type: 'iteration',
```
Comment (kunagpal) [inline lib/index.js]: Top level `const`

File: lib/index.js [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+                model: {
+                    $self: ['iterations', null]
```
Comment (kunagpal) [inline lib/index.js]: See comment above.

File: lib/index.js [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                cursor: data.cursor,
+                context: {
+                    ref: data.cursor.ref,
+                    httpRequestId: data.cursor.httpRequestId,
+                    scriptId: data.cursor.scriptId,
+                    execution: data.cursor.execution
+                },
+                run: {
+                    $self: ['run', runId]
```
Comment (kunagpal) [inline lib/index.js]: Here too.

File: lib/index.js [hunk 14]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                error: err && { $self: ['errors', err.id] },
```
Comment (kunagpal) [inline lib/index.js]: And here as well.

File: lib/index.js [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+                timings: {
+                    started: iterationStartTime,
+                    completed: Date.now()
+                }
+            });
+        });
+
+        newman.on('beforeItem', function () {
+            itemStartTime = Date.now();
```
Comment (kunagpal) [inline lib/index.js]: Comment explaining intent.

File: lib/index.js [hunk 16]
```diff
@@ -0,0 +1,437 @@
...
+                errors[err.id] = err;
+            }
+
+            const executionId = uuid();
+
+            executions.push({
+                id: executionId,
+                index: executions.length,
+                branch: 0,
+                type: 'item',
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 17]
```diff
@@ -0,0 +1,437 @@
...
+                model: {
+                    $self: ['items', data.item.id]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 18]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                cursor: data.cursor,
+                context: {
+                    ref: data.cursor.ref,
+                    httpRequestId: data.cursor.httpRequestId,
+                    scriptId: data.cursor.scriptId,
+                    execution: data.cursor.execution
+                },
+                run: {
+                    $self: ['run', runId]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 19]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                error: err && { $self: ['errors', err.id] },
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 20]
```diff
@@ -0,0 +1,437 @@
...
+            !items[data.item.id] && (items[data.item.id] = _items[data.item.id]);
+        });
+
+        newman.on('beforeScript', function () {
+            scriptStartTime = Date.now();
+        });
+
+        newman.on('script', function (err, data) {
+            if (err) {
+                !(err instanceof SerialisedError) && (err = new SerialisedError(err, true));
```
Comment (kunagpal) [inline lib/index.js]: See comment above about changing types on the fly. ✈️

File: lib/index.js [hunk 21]
```diff
@@ -0,0 +1,437 @@
...
+                errors[err.id] = err;
+            }
+
+            const executionId = uuid();
+
+            executions.push({
+                id: executionId,
+                index: executions.length,
+                branch: 0,
+                type: 'script',
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 22]
```diff
@@ -0,0 +1,437 @@
...
+                model: {
+                    $self: ['scripts', data.execution.id]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 23]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                cursor: data.cursor,
+                context: {
+                    ref: data.cursor.ref,
+                    httpRequestId: data.cursor.httpRequestId,
+                    scriptId: data.cursor.scriptId,
+                    execution: data.cursor.execution
+                },
+                run: {
+                    $self: ['run', runId]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 24]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                error: err && { $self: ['errors', err.id] },
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 25]
```diff
@@ -0,0 +1,437 @@
...
+                    started: scriptStartTime,
+                    completed: Date.now()
+                }
+            });
+
+            scripts[data.execution.id] = {
+                id: data.execution.id,
+                listen: data.event.listen,
+                execution: {
+                    $self: ['executions', executionId]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 26]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                event: {
+                    $self: ['events', data.script.id]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 27]
```diff
@@ -0,0 +1,437 @@
...
+                }
+            };
+
+            // Add the event to known events list
+            !events[data.script.id] && (events[data.script.id] = _events[data.script.id]);
+        });
+
+        newman.on('beforeRequest', function () {
+            requestStartTime = Date.now();
```
Comment (kunagpal) [inline lib/index.js]: Comment explaining intent.

File: lib/index.js [hunk 28]
```diff
@@ -0,0 +1,437 @@
...
+        });
+
+        newman.on('request', function (err, data) {
+            if (err) {
+                !(err instanceof SerialisedError) && (err = new SerialisedError(err, true));
```
Comment (kunagpal) [inline lib/index.js]: See comment above about changing types on the fly.

File: lib/index.js [hunk 29]
```diff
@@ -0,0 +1,437 @@
...
+
+            const executionId = uuid();
+
+            executions.push({
+                id: executionId,
+                index: executions.length,
+                branch: 0,
+                type: 'request',
+                model: {
+                    $self: ['requests', data.cursor.httpRequestId]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 30]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                cursor: data.cursor,
+                context: {
+                    ref: data.cursor.ref,
+                    httpRequestId: data.cursor.httpRequestId,
+                    scriptId: data.cursor.scriptId,
+                    execution: data.cursor.execution
+                },
+                run: {
+                    $self: ['run', runId]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 31]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                error: err && { $self: ['errors', err.id] },
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 32]
```diff
@@ -0,0 +1,437 @@
...
+                timings: {
+                    started: requestStartTime,
+                    completed: Date.now()
+                }
+            });
+
+            requests[data.cursor.httpRequestId] = {
+                id: data.cursor.httpRequestId,
+                execution: {
+                    $self: ['executions', executionId]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 33]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                response: {
+                    $self: ['responses', data.response && data.response.id]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 34]
```diff
@@ -0,0 +1,437 @@
...
+                url: data.request.url.toString(),
+                method: data.request.method
+            };
+
+            // If there was no response then do not add it
+            if (!data.response) { return; }
+
+            responses[data.response.id] = {
+                request: {
+                    $self: ['requests', data.cursor.httpRequestId]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 35]
```diff
@@ -0,0 +1,437 @@
...
+
+        newman.on('assertion', function (err, data) {
+            const executionId = uuid(),
+                assertionId = uuid();
+
+            executions.push({
+                id: executionId,
+                index: executions.length,
+                branch: 0,
+                type: 'assertion',
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 36]
```diff
@@ -0,0 +1,437 @@
...
+                model: {
+                    $self: ['assertions', assertionId]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 37]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                cursor: data.cursor,
+                context: {
+                    ref: data.cursor.ref,
+                    httpRequestId: data.cursor.httpRequestId,
+                    scriptId: data.cursor.scriptId,
+                    execution: data.cursor.execution
+                },
+                run: {
+                    $self: ['run', runId]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 38]
```diff
@@ -0,0 +1,437 @@
...
+                    started: Date.now(),
+                    completed: Date.now()
+                }
+            });
+
+            assertions[assertionId] = {
+                id: assertionId,
+                name: data.assertion,
+                script: {
+                    $self: ['scripts', data.cursor.execution]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 39]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                skipped: data.skipped,
+                error: err
+            };
+        });
+
+        newman.on('console', function (err, data) {
+            if (err) {
+                !(err instanceof SerialisedError) && (err = new SerialisedError(err, true));
```
Comment (kunagpal) [inline lib/index.js]: Types being changed on the fly.

File: lib/index.js [hunk 40]
```diff
@@ -0,0 +1,437 @@
...
+            }
+
+            const consoleId = uuid(),
+                executionId = uuid();
+
+            executions.push({
+                id: executionId,
+                index: executions.length,
+                branch: 0,
+                type: 'console',
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 41]
```diff
@@ -0,0 +1,437 @@
...
+                model: {
+                    $self: ['consoles', consoleId]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 42]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                cursor: data.cursor,
+                context: {
+                    ref: data.cursor.ref,
+                    httpRequestId: data.cursor.httpRequestId,
+                    scriptId: data.cursor.scriptId,
+                    execution: data.cursor.execution
+                },
+                run: {
+                    $self: ['run', runId]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 43]
```diff
@@ -0,0 +1,437 @@
...
+                },
+                error: err && { $self: ['errors', err.id] },
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: lib/index.js [hunk 44]
```diff
@@ -0,0 +1,437 @@
...
+                timings: {
+                    started: Date.now(),
+                    completed: Date.now()
+                }
+            });
+
+            consoles[consoleId] = {
+                id: consoleId,
+                script: {
+                    $self: ['scripts', data.cursor.execution]
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: npm/test-integration.js [hunk 45]
```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env node
+/* global test, mkdir, rm, exit */
+require('shelljs/global');
```
Comment (kunagpal) [inline npm/test-integration.js]: Avoid using shelljs as a global.

File: npm/test-integration.js [hunk 46]
```diff
@@ -0,0 +1,81 @@
...
+require('colors');
+
+var fs = require('fs'),
+    path = require('path'),
+    Mocha = require('mocha'),
+    expect = require('chai').expect,
+    newman = require('../node_modules/newman'),
```
Comment (kunagpal) [inline npm/test-integration.js]: Update this to attain parity with the html reporter.

File: test/fixtures/failed-request.json [hunk 47]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (kunagpal) [inline test/fixtures/failed-request.json]: All JSON files use 2 spaced indent 😈

File: test/fixtures/failed-request.json [hunk 48]
```diff
@@ -0,0 +1,18 @@
...
+		"schema": "https://schema.getpostman.com/json/collection/v2.0.0/collection.json"
+	},
+	"item": [
+		{
+			"name": "Failed request",
+			"request": {
+				"url": "https://123.random.z/get",
+				"method": "GET"
+			}
+    }
```
Comment (kunagpal) [inline test/fixtures/failed-request.json]: ?

File: test/fixtures/sample-collection.json [hunk 49]
```diff
@@ -0,0 +1,627 @@
...
+	],
+	"variable": [
+		{
+			"id": "c8532587-e0da-4fa0-821d-a032c233bd83",
+			"key": "host",
+			"value": "https://postman-echo.com",
+			"type": "string"
+		}
+	]
+}
```
Comment (kunagpal) [inline test/fixtures/sample-collection.json]: Newline needed at the end of this file.

File: test/integration/library.test.js [hunk 50]
```diff
@@ -0,0 +1,119 @@
+const nock = require('nock'),
+    Ajv = require('ajv');
+
+describe('Newman Library', function () {
+    before(function (done) {
+        var ajv = new Ajv(),
+            validate = ajv.compile(require('../../examples/schema.json'));
```
Comment (kunagpal) [inline test/integration/library.test.js]: `require`s go to the top level.

File: lib/index.js [hunk 51]
```diff
@@ -114,141 +197,78 @@ const os = require('os'),
...
-                error: err && { $self: ['errors', err.id] },
-                timings: {
-                    started: iterationStartTime,
-                    completed: Date.now()
-                }
-            });
+            addExecution(err,
+                data,
+                executions,
+                EXECUTION_TYPES.iteration,
```
Comment (kunagpal) [inline lib/index.js]: Check if it is possible to infer the execution type from the model type or vice versa. This helps remove one argument from the signature.

File: test/fixtures/failed-request.json (hunk 47 above)
Comment (codenirvana) [reply test/fixtures/failed-request.json]: Done

File: README.md [hunk 52]
```diff
@@ -41,10 +41,20 @@ newman.run({
...
+
+            // post execution summary to remote endpoint
+            // request.js(https://github.com/request/request) options object
+            api: {
                 url: 'http://localhost:3000',
                 method: 'POST'
+            },
+
+            // custom handling of execution summary
+            callback: function(summary) {
```
Comment (kunagpal) [inline README.md]: Space before `(`

File: lib/index.js [hunk 53]
```diff
@@ -393,21 +394,25 @@ const os = require('os'),
...
+                    console.info('newman-reporter-remote:\n  responseStatus: ' + response.statusCode +
+                        '\n  responseBody: ', body);
+                });
+            }
 
-                console.info('newman-reporter-remote:\n  responseStatus: ' + response.statusCode +
-                    '\n  responseBody: ', body);
-            });
+            // callback for custom handling of execution summary
+            if (typeof options.callback === 'function') {
```
Comment (kunagpal) [inline lib/index.js]: `const`

File: examples/schema.json [hunk 54]
```diff
@@ -0,0 +1,789 @@
+{
+  "$id": "https://schema.getpostman.com/json/execution/v1.0.0/execution.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "type": "object",
+  "title": "ERS Schema",
```
Comment (kunagpal) [inline examples/schema.json]: ?

File: examples/schema.json [hunk 55]
```diff
@@ -0,0 +1,789 @@
...
+            "$id": "#/properties/executions/items/properties/index",
+            "type": "integer",
+            "title": "Index of the executions",
+            "examples": [
+              0
+            ]
+          },
+          "branch": {
+            "$id": "#/properties/executions/items/properties/branch",
+            "type": "integer",
```
Comment (kunagpal) [inline examples/schema.json]: Non negative value.

File: examples/schema.json [hunk 56]
```diff
@@ -0,0 +1,789 @@
...
+          "timings": {
+            "$id": "#/properties/executions/items/properties/timings",
+            "type": "object",
+            "title": "Execution timings",
+            "required": [
+              "started",
+              "completed"
+            ],
+            "properties": {
+              "started": {
```
Comment (kunagpal) [inline examples/schema.json]: Both should be non-negative.

File: examples/schema.json [hunk 57]
```diff
@@ -0,0 +1,789 @@
...
+                "$id": "#/properties/executions/items/properties/timings/properties/started",
+                "type": "integer",
+                "title": "Execution start time",
```
Comment (kunagpal) [inline examples/schema.json]: Both these times are absolute.

File: examples/schema.json [hunk 58]
```diff
@@ -0,0 +1,789 @@
...
+            },
+            "status": {
+              "type": "string",
+              "title": "Response status",
+              "examples": [
+                "OK"
+              ]
+            },
+            "code": {
+              "type": "integer",
```
Comment (kunagpal) [inline examples/schema.json]: Non negative, all of them.

File: lib/index.js [hunk 59]
```diff
@@ -0,0 +1,421 @@
...
+                    }
+
+                    console.info('newman-reporter-remote:\n  responseStatus: ' + response.statusCode +
+                        '\n  responseBody: ', body);
+                });
+            }
+
+            // callback for custom handling of execution summary
+            if (typeof options.callback === FUNCTION) {
+                options.callback(summary);
```
Comment (kunagpal) [inline lib/index.js]: `null` as a first argument.

File: test/integration/library.test.js [hunk 60]
```diff
@@ -0,0 +1,66 @@
+const Ajv = require('ajv'),
+    EXECUTION_SCHEMA = require('../../examples/schema.json');
+
+describe('Newman Library', function () {
+    const Validate = new Ajv().compile(EXECUTION_SCHEMA);
```
Comment (kunagpal) [inline test/integration/library.test.js]: `validate`

File: test/system/travis-yml.test.js [hunk 61]
```diff
@@ -0,0 +1,28 @@
+describe('travis.yml', function () {
+    var fs = require('fs'),
```
Comment (kunagpal) [inline test/system/travis-yml.test.js]: Shift to the outer scope.

**Pull request merged**